### PR TITLE
One-Way Windows

### PR DESCRIPTION
**Pull request by Shadowmech88**
Revival of #10429

A window can now be turned into a one-way window by using a sheet of plastic on it.
The plastic can be pried off with a crowbar, turning them back into regular windows.
One-way windows obstruct a player's view from one side, but allow them to see through from the other.

One-way windows block night vision goggles and thermals, but not X-ray vision.
They also don't prevent right-clicking to see the contents of a tile. This cannot be fixed.

one_way_window

The window's dir affects the directions from which it blocks vision, even for full windows. When building one-way windows, face the window toward the direction from which you want it to be transparent.
Full windows have a new verb that allows players to manually set their direction, so that they can access the window's diagonal directions, since this PR gives diagonal windows an actual gameplay difference from cardinal windows. Window smoothing doesn't seem to be affected.

For mapping purposes, regular windows can be made into one-way windows by toggling their one_way var.

I'll add these to the Interrogation room once they're merged.

:cl:
 * rscadd: Added one-way windows. Construct by using a plastic sheet on a window.

### CODE REVIEW

**Comment by Shadowmech88** [comment]
I tested it locally and it just werked, but it also worked last time yet caused problems when it was merged.
This is probably going to need to be tested publicly on the PR test server or something to make sure it's actually working.

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: A better version of this optimization is already built into `update_parallax_values()`. Adding it again here will only worsen performance.

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: Loop directly through `view()` rather than using a variable. It seems counter-intuitive, but BYOND has internal optimizations for loops through `view()`

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: Use the client's `view` rather than `world.view`, and remove the `get_holder_at_turf_level()`. `view()` probably internally gets the turf of the second arg. If it doesn't, use `get_turf()` instead.
Except not here; see below

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: Again, it's better to loop directly through `view()`
I assume there's a tipping point at which the number of calls to `view()` outweighs the optimization, but I'm quite confident that tipping point is at way more than 2

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: `get_dir()` also probably internally gets the turf. And again, if it doesn't, use `get_turf()`

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: Again, client's `view`. And again, don't assign a variable. Just loop directly through `view()`. And also no `get_holder_at_turf_level()`

File: code/modules/client/client procs.dm
Comment (Exxion) [inline code/modules/client/client procs.dm]: Define

File: code/modules/mob/mob.dm
Comment (Exxion) [inline code/modules/mob/mob.dm]: This check is unnecessary.

**Review by Exxion — DISMISSED** [review]

File: code/modules/client/client procs.dm
Comment (Shadowmech88) [reply code/modules/client/client procs.dm]: The variable must be assigned because I need to compare the tiles visible before making all the tiles behind the windows opaque with the tiles visible after making all the windows opaque, in order to get a list of the tiles that *would* be obscured if the windows were opaque, so that I can manually obscure them afterward.

Comment (Exxion) [reply code/modules/client/client procs.dm]: You're right
Ergh

**Comment by Durgen** [comment]
hopefully these will get mapped into the brig, and people will actually start using interview rooms.

can you hear someone talking on the other side of them?

**Comment by Shadowmech88** [comment]
You can't hear someone talking through these if you're on the side where you can't see the person.
However, devices like intercoms and tape recorders *can*, unfortunately. I believe the performance impact of trying to remedy that would be too large.

**Comment by NigglyWiggly** [comment]
Don't tinted windows already have this functionality or am I forgetting things?

**Comment by Shadowmech88** [comment]
Tinted windows are just opaque. They're walls that look like windows.

**Comment by Exxion** [comment]
Honestly the performance impact of it even without making it apply to those might be too large
But I guess we'll see

File: code/game/objects/structures/window.dm
Comment (ComicIronic) [inline code/game/objects/structures/window.dm]: Use `getFromPool` for stacks.

File: code/modules/client/client procs.dm
Comment (ComicIronic) [inline code/modules/client/client procs.dm]: `break` should not appear outside loops.

**Comment by ComicIronic** [comment]
I'm still kind of against this. It doesn't work fully without a major performance hit, and it will still break in many ways because Lummox will never provide an eye-movement hook to attach the logic to properly, so I expect many reports of "one-way windows bypassed by lockers". It's going to introduce bugs, not because you made it badly, but just because it doesn't work, and it can't work.

**Comment by Shadowmech88** [comment]
>so I expect many reports of "one-way windows bypassed by lockers"

Parallax doesn't stop working while inside a locker, why can't we put a call to the proc that updates the window view inside whatever code does that?
Or is that what you mean by "doesn't work fully without a major performance hit"?

**Comment by ComicIronic** [comment]
My mistake, parallax now works inside of lockers because we do some amount of mob recursion, I believe. Does the window darkness now catch clicks or not?

**Comment by Shadowmech88** [comment]
The window darkness catches regular clicks. Right clicks still function as though the darkness isn't there.

**Comment by Intigracy** [comment]
I'll merge this Monday, hopefully it doesn't have as bad of a time as last time.

**Comment by Shadowmech88** [comment]
Before we merge it for real, I'd like to testmerge it on the PR test server and make sure it works there.

**Comment by Exxion** [comment]
The problem before was probably number of people. I'm not sure that would help.

**Comment by Shadowmech88** [comment]
There weren't even any one-way windows on the map during the initial merge, to my knowledge.
From what I can see, the only performance impact this would have without any one-way windows actually being in-game is a single additional `view()` check per client per client hook update.

Would that really cause catastrophic lag?

**Comment by Exxion** [comment]
Probably not, but I've been surprised before
But we've made some optimizations since last time, too. Would there have been more last time?

**Comment by Shadowmech88** [comment]
Possibly. One major difference is the lack of direct `view()` calls in the old PR because I didn't know that `view()` was internally optimized.

**Comment by Intigracy** [comment]
Not sure how to go about getting this on the test server.

**Comment by Shadowmech88** [comment]
@Intigracy 
In the coding discord, collaborators can put PRs onto the PR test server by pinging the MoMMI bot with `testmerge PRnumber`
IE `@MoMMI testmerge 14777`

**Comment by Shadowmech88** [comment]
This was loaded onto the PR test server and it crashed after about 15 minutes.
There don't seem to be any logs or other useful information regarding exactly *why* it crashed, but don't merge this just yet, just to be safe.

**Comment by Shadowmech88** [comment]
Another test on the PR test server has been going strong for two hours, so maybe the previous crash was just a fluke.
Can't know for sure though, because not hardly enough people connected to stress test it.

**Comment by Shadowmech88** [comment]
It turns out N3X15 was using the PR test server to also host a Factorio server, it looks like the crash was due to that eating up all the resources.

**Comment by Intigracy** [comment]
Ping me on discord if this needs reverting.

**Pull request merged**
